multimap/src/main/java/org/infinispan/multimap/impl/ListBucket.java

```diff
@@ -87,13 +87,62 @@ public String toString() {
    }
 
    public ListBucket<V> offer(V value, boolean first) {
-      Deque<V> newBucket = new ArrayDeque<>(values);
       if (first) {
-         newBucket.offerFirst(value);
+         values.offerFirst(value);
       } else {
-         newBucket.offerLast(value);
+         values.offerLast(value);
       }
 
+      return new ListBucket<>(values);
+   }
+
+   public ListBucket<V> set(long index, V value) {
+      if ((index >= 0 && (values.size()-1 < index)) || (index < 0 && (values.size() + index < 0))) {
+        return null;
+      }
+
+      // set head
+      if (index == 0) {
+         values.pollFirst();
+         values.offerFirst(value);
+         return new ListBucket<>(values);
+      }
+
+      // set tail
+      if (index == -1) {
```

Comment:
```suggestion
      if (index == -1 || index == values.size -1) {
```
Just a small optimization if they ended up picking the last node.